spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/RoiAlign.scala

```diff
@@ -109,8 +111,180 @@ class RoiAlign[T: ClassTag] (
     output
   }
 
-  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
-    throw new UnsupportedOperationException("Not support backward propagation")
+
+  private def bilinearInterpolateGradient(height: Int, width: Int, y: Float, x: Float)
+    : (Float, Float, Float, Float, Int, Int, Int, Int) = {
+    var w1: Float = 0.0f
+    var w2: Float = 0.0f
+    var w3: Float = 0.0f
+    var w4: Float = 0.0f
+    var x_low : Int = 0
+    var x_high: Int = 0
+    var y_low: Int = 0
+    var y_high: Int = 0
+
+    // deal with cases that inverse elements are out of feature map boundary
+    if (y < -1.0 || y > height || x < -1.0 || x > width) {
+      // empty
+      return (w1, w2, w3, w4, x_low, x_high, y_low, y_high)
+    }
+
+    var realY = if (y <= 0) 0 else y
+    var realX = if (x <= 0) 0 else x
+    y_low = realY.toInt
+    x_low = realX.toInt
+
+    if (y_low >= height - 1) {
+      y_high = height - 1
+      y_low = height - 1
+      realY = y_low
+    } else y_high = y_low + 1
+
+    if (x_low >= width - 1) {
+      x_high = width - 1
+      x_low = width - 1
+      realX = x_low
+    } else x_high = x_low + 1
+
+    val ly = realY - y_low
+    val lx = realX - x_low
+    val hy = 1.0 - ly
+    val hx = 1.0 - lx
+
+    w1 = (hy * hx).toFloat
+    w2 = (hy * lx).toFloat
+    w3 = (ly * hx).toFloat
+    w4 = (ly * lx).toFloat
+
+    return (w1, w2, w3, w4, x_low, x_high, y_low, y_high)
+  }
+
+  private def roiAlignBackward(
+    nums: Int,
+    gradOutputArr: Array[T],
+    gradInputArr: Array[T],
+    gradInputOffset: Int,
+    rois: Array[Float],
+    channels: Int,
+    height: Int,
+    width: Int,
+    pooled_height: Int,
+    pooled_width: Int,
+    sampling_ratio : Int,
+    n_stride : Int,
+    c_stride : Int,
+    h_stride : Int,
+    w_stride : Int,
+    spatial_scale: Float) {
+    val roi_cols = 4
+    for (index <- 0 until nums) {
+      val pw = index % pooled_width
+      val ph = (index / pooled_width) % pooled_height
+      val c = (index / pooled_width / pooled_height) % channels
+      val n = index / pooled_width / pooled_height / channels
+      val offset_rois = n * roi_cols
+
+      val offset = if (aligned) 0.5f else 0.0f
+      val roi_start_w = rois(offset_rois) * spatial_scale - offset
+      val roi_start_h = rois(offset_rois + 1) * spatial_scale - offset
+      val roi_end_w = rois(offset_rois + 2) * spatial_scale - offset
+      val roi_end_h = rois(offset_rois + 3) * spatial_scale - offset
+
+      var roi_width = roi_end_w - roi_start_w
+      var roi_height = roi_end_h - roi_start_h
+
+      if (aligned) {
+        require(roi_width >= 0 && roi_height >= 0,
+          s"ROIs in ROIAlignRotated do not have non-negative size!" +
+            s"But get ${roi_height} ${roi_width}")
+      } else {
+        roi_width = math.max(roi_width, 1.0f)
```

Comment:
Why this part is skipped if aligned?

Reply:
for aligned case, we allow roi height/width less than 1.